### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by ThomasMarwitzQC**
Refactoring changes include:
- [x] Include Python 3.12 in CI
- [x] Migrate `setup.cfg` and `setup.py` into `pyproject.toml`
- [x] Port to ruff
- [x] Include pre-commit autoupdate workflow  
- [x] Solved Issue #44 (as it relates to the refactoring) 
- [x] Solution for `google_auth` step. 
- [ ] Assign new tag `1.8.2`

### CODE REVIEW

**Comment by ThomasMarwitzQC** [comment]
Added ruff formatter @xhochy

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -1,21 +1,14 @@
 repos:
- - repo: https://github.com/Quantco/pre-commit-mirrors-black
-   rev: 23.7.0
+ # Run ruff first because autofix behaviour is enabled
+ - repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (xhochy) [inline .pre-commit-config.yaml:3]: Please use the QuantCo-provided conda-based version of this.

**Review by xhochy — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,21 +1,13 @@
...
-     - id: black-conda
-       args:
-         - --safe
-         - --target-version=py38
- - repo: https://github.com/Quantco/pre-commit-mirrors-flake8
-   rev: 6.1.0
+    - id: ruff-conda
+ - repo: https://github.com/Quantco/pre-commit-mirrors-ruff
+   rev: "0.1.3"
    hooks:
```
Comment (xhochy) [inline .pre-commit-config.yaml:9]: Can we have a single ruff section?
```suggestion
```

**Review by xhochy — APPROVED** [review]
Can you change the title and "undraft"? Then this can go in.

**Pull request merged**